src/lex.rs

```diff
@@ -108,78 +106,88 @@ impl<'a> Lexer<'a> {
 
   /// Gets the next word in the input, as well as whether it's on
   /// a different line than the last word we got.
-  fn next_word(&mut self) -> Option<Vec<u8>> {
-    let mut ret: Vec<u8> = Vec::new();
-
+  fn next_word(&mut self) -> Option<&'a [u8]> {
     self.skip_whitespace_except_newline();
+    self.skip_comment();
 
-    loop {
-      match self.peek() {
-        None => break,
-        Some(c) => {
-          if c == b'#' {
-            assert!(self.skip_comment());
-            self.skip_whitespace_except_newline();
-          } else if is_whitespace(c) {
-            if c == b'\n' && ret.is_empty() {
-              ret.push(c);
-              self.advance();
-            }
-            break;
-          } else {
-            ret.push(c);
-            self.advance();
-          }
+    let start_ptr = self.read_pos;
+
+    match self.peek() {
+      Some(b'\n') => {
+        self.advance();
+        Some(&self.bytes[start_ptr..self.read_pos]) // newline
+      }
+      Some(_) => {
+        if self.skip_unless(|c| is_whitespace(c) || c == b'#') {
+          Some(&self.bytes[start_ptr..self.read_pos])
+        } else {
+          None
         }
       }
-    }
-
-    if ret.is_empty() {
-      debug_assert_eq!(self.peek(), None);
-      None
-    } else {
-      Some(ret)
+      None => None,
     }
   }
 }
 
-impl<'a> Iterator for Lexer<'a> {
-  type Item = String;
+#[derive(Clone)]
+pub struct PeekableLexer<'a> {
+  inner: Lexer<'a>,
+  peeked: Option<Option<&'a str>>,
+}
 
-  fn next(&mut self) -> Option<String> {
-    self.next_word().map(|buf| {
-      match String::from_utf8(buf) {
-        Ok(s) => s,
-        Err(_) => panic!("Lex error: Invalid utf8 on line {}.", self.current_line_number),
-      }
-    })
+impl<'a> PeekableLexer<'a> {
+  pub fn new(lexer: Lexer<'a>) -> Self {
+    Self {
+      inner: lexer,
+      peeked: None,
+    }
+  }
+  pub fn next_str(&mut self) -> Option<&'a str> {
+    match self.peeked.take() {
+      Some(v) => v,
+      None => self
+        .inner
+        .next_word()
+        .map(|buf| unsafe { str::from_utf8_unchecked(buf) }),
```

Comment:
I know that, but the buffer is always coming from either String or str source, so it's already validated. No need to do extra checks here, as it visibly impacts performance.

Reply:
Any special utf8 sequences involves msb being set in the bytes, which mean non-ascii characters. All operations that are done in the parser always consumes such sequences entirely, because we only split words on ascii whitespaces and newlines. This effectively guarantees that a word will always be a valid utf8 substring.